### PR TITLE
Enforce md5sum checks for object retention APIs

### PR DESCRIPTION
**Pull request by poornas**
this PR enforces md5sum verification for following
API's to be compatible with AWS S3 spec
 - PutObjectRetention
 - PutObjectLegalHold
## How to test this PR?
using aws cli

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by harshavardhana** [comment]
Did we decide to do this?

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#5212df4b-e58f-4b02-a0c9-1c9ac760fd94) |
| mint-gateway-azure.sh | :x: [more...](#6834c679-4d73-4e6e-9b6f-2f07e7bcbdb8) |
<a name="6834c679-4d73-4e6e-9b6f-2f07e7bcbdb8"></a>
### 9030-7b624d3/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:30502
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1a0b3384e80c:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 4 minutes and 54 seconds
(4/14) Running aws-sdk-ruby tests ... done in 17 seconds
(5/14) Running awscli tests ... FAILED in 24 seconds
{
  "name": "awscli",
  "duration": 2199,
  "function": "aws --endpoint-url http://72.28.97.52:30502 s3api put-object --body /mint/data/datafile-0-b --bucket \nAn error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it. --key datafile-0-b\n",
  "status": "FAIL",
  "error": "usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]To see help text, you can run:  aws help  aws <command> help  aws <command> <subcommand> helpUnknown options: occurred, (BucketAlreadyOwnedByYou), when, calling, the, CreateBucket, operation:, Your, previous, request, to, create, the, named, bucket, succeeded, and, you, already, own, it., error"
}

Executed 4 out of 14 tests successfully.
```
<a name="5212df4b-e58f-4b02-a0c9-1c9ac760fd94"></a>
### 9030-7b624d3/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:30563
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9ef6debd7daf:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 45 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 21 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 32 seconds
(8/14) Running minio-dotnet tests ... done in 41 seconds
(9/14) Running minio-go tests ... done in 10 minutes and 0 seconds
(10/14) Running minio-java tests ... done in 59 seconds
(11/14) Running minio-js tests ... FAILED in 10 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 13,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:85:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:155:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:79:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}

Executed 10 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by poornas** [comment]
> Did we decide to do this?

I believe so as per @kannappanr

**Comment by harshavardhana** [comment]
@poornas this PR is not needed and frivolous because we already have a generalized implementation in auth-handler.go which is implemented for all APIs 

With this PR we will be doing double hash

**Comment by poornas** [comment]
> @poornas this PR is not needed and frivolous because we already have a generalized implementation in auth-handler.go which is implemented for all APIs
> 
> With this PR we will be doing double hash

 Retention APIs were not using the auth-handler checkRequestAuthType function that does  Content-MD5 verification. Fixing my patch to utilize this generic implementation

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#97cfe561-9640-4de3-b6b3-8cb0c148411b) |
<a name="97cfe561-9640-4de3-b6b3-8cb0c148411b"></a>
### 9030-6384637/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:30545
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 035492bfca7f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 4 seconds
(4/14) Running aws-sdk-ruby tests ... done in 17 seconds
(5/14) Running awscli tests ... done in 3 minutes and 1 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 6 minutes and 40 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 34 seconds
(9/14) Running minio-go tests ... FAILED in 16 minutes and 55 seconds
{
  "args": {
    "bucketName": "minio-go-test-r6utj4k0nkw19xci",
    "objectName": "yzu11xmz366zdqefuy9ex6nc159isy"
  },
  "duration": 112152,
  "error": "Get http://72.28.97.52:30545/minio-go-test-r6utj4k0nkw19xci/yzu11xmz366zdqefuy9ex6nc159isy: Connection closed by foreign host http://72.28.97.52:30545/minio-go-test-r6utj4k0nkw19xci/yzu11xmz366zdqefuy9ex6nc159isy. Retry again.",
  "function": "GetObject(bucketName, objectName)",
  "message": "ReadAt failed",
  "name": "minio-go: testGetObjectReadAtWhenEOFWasReached",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#a061f981-bacb-47a5-bf5b-0663e30dc026) |
| mint-gateway-nas.sh | :x: [more...](#aea679e0-dcba-48d8-b554-bb18eaa2ba02) |
<a name="aea679e0-dcba-48d8-b554-bb18eaa2ba02"></a>
### 9030-e4d2481/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31402
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6738864ea605:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 2 minutes and 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-kkalc4rlncmifi",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120368,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:31402/aws-sdk-go-test-kkalc4rlncmifi: dial tcp 72.28.97.55:31402: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="a061f981-bacb-47a5-bf5b-0663e30dc026"></a>
### 9030-e4d2481/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30022
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2afe0dfb192f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 12 seconds
(4/14) Running aws-sdk-ruby tests ... done in 17 seconds
(5/14) Running awscli tests ... FAILED in 1 minutes and 27 seconds
{
  "name": "awscli",
  "duration": 1536,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -2806,18 +2788,7 @@ func (api objectAPIHandlers) PutObjectRetentionHandler(w http.ResponseWriter, r
...
-		data, err := xml.Marshal(objRetention)
-		if err != nil {
-			writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
-			return
-		}
-		if hex.EncodeToString(md5Bytes) != getMD5Hash(data) {
-			writeErrorResponse(ctx, w, errorCodes.ToAPIErr(ErrInvalidDigest), r.URL, guessIsBrowserReq(r))
-			return
-		}
-	}
```
Comment (nitisht) [inline cmd/object-handlers.go:2823]: Probably I am missing something, but where does the `md5` validation happen now?

Comment (harshavardhana) [reply cmd/object-handlers.go:2823]: @nitisht in checkRequestAuthType()

**Review by nitisht — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#ebb055ea-8ede-49b3-97a6-f9676c486011) |
| mint-gateway-azure.sh | :x: [more...](#a107c334-bbb1-4cb5-a80c-e0d1519e759e) |
<a name="a107c334-bbb1-4cb5-a80c-e0d1519e759e"></a>
### 9030-b72badc/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31759
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e6bc635ad82e:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 35 seconds
(4/14) Running aws-sdk-ruby tests ... done in 19 seconds
(5/14) Running awscli tests ... FAILED in 52 seconds
{
  "name": "awscli",
  "duration": 1531,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (BucketAlreadyOwnedByYou) when calling the CreateBucket operation: Your previous request to create the named bucket succeeded and you already own it."
}

Executed 4 out of 14 tests successfully.
```
<a name="ebb055ea-8ede-49b3-97a6-f9676c486011"></a>
### 9030-b72badc/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32004
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a155af76b89b:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 2 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-4y53opa9fudbu2",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 120298,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.55:32004/aws-sdk-go-test-4y53opa9fudbu2: dial tcp 72.28.97.55:32004: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by nitisht** [comment]
Blocked for merging till clients catch up and send md5 sums with relevant requests

**Comment by harshavardhana** [comment]
@nitisht this md5sum has nothing to do with PutObject md5sum requirement. Here the validation is about the XML body to be enforced with md5sum.

**Pull request merged**